Unexpected attempted differentiation of non-differentiation function error

Reproducible example:

```cpp
__attribute__((annotate("non_differentiable")))
void do_nothing(double u, double v) {}

double fn(double u, double v) { 
  do_nothing(u, v);
  return u + v; 
}

int main() {
  auto d_f1 = clad::gradient(fn);
}
```

Output:
```
 error: attempted differentiation of function 'do_nothing', which is marked as non-differentiable
  do_nothing(u, v);
```

In this example, Clad is requested to differentiate `fn` and this function internally calls a non-differentiable function `do_nothing`. Clad should not try to differentiate `do_nothing` as it is a non-differentiable function.

> However, when I compile and run the following source code on my system, it successfully generates the gradient without any errors.

The code you provided is somewhat messy. After organizing it, I compiled your code using LLVM 18.1.3 but couldn't reproduce your result. My result is the same as @parth-07's:

```
temp.cpp:7:3: error: attempted differentiation of function 'do_nothing', which is marked as non-differentiable  
    7 |   do_nothing(u, v);  
      |   ^ 
```

--- a/lib/Differentiator/DiffPlanner.cpp
+++ b/lib/Differentiator/DiffPlanner.cpp
@@ -1040,6 +1040,12 @@ namespace clad {
       }
       m_TopMostReq = &request;
     } else {
+      // If the function contains annotation of non_differentiable, then Clad
+      // should not produce any derivative expression for that function call,
+      // and the function call in the primal should be used as it is.
+      if (clad::utils::hasNonDifferentiableAttribute(E))
+        return true;
+
       // Don't build propagators for calls that do not contribute in
       // differentiable way to the result.
       if (!isa<CXXMemberCallExpr>(E) && !isa<CXXOperatorCallExpr>(E) &&

--- a/test/ForwardMode/NonDifferentiable.C
+++ b/test/ForwardMode/NonDifferentiable.C
@@ -67,6 +67,15 @@ double fn_s2_operator(double i, double j) {
   return (obj1 + obj2).mem_fn(i, j);
 }
 
+non_differentiable
+double fn_non_diff(double i, double j) {
+  return i * j;
+}
+
+double fn_non_diff_call(double i, double j) {
+  return fn_non_diff(i, j) + i * j;
+}
+
 #define INIT_EXPR(classname)                                                   \
   classname expr_1(2, 3);                                                      \
   classname expr_2(3, 5);
@@ -108,6 +117,8 @@ int main() {
   TEST_FUNC(fn_s2_field, 3, 5) // CHECK-EXEC: 5.00
 
   TEST_FUNC(fn_s2_operator, 3, 5) // CHECK-EXEC: 0.00
+
+  TEST_FUNC(fn_non_diff_call, 3, 5) // CHECK-EXEC: 5.00
 
   // CHECK: double mem_fn_1_darg0(double i, double j) {
   // CHECK-NEXT:     double _d_i = 1;
@@ -207,6 +218,12 @@ int main() {
   // CHECK-NEXT:     return 0;
   // CHECK-NEXT: }
 
+  // CHECK: double fn_non_diff_call_darg0(double i, double j) {
+  // CHECK-NEXT:     double _d_i = 1;
+  // CHECK-NEXT:     double _d_j = 0;
+  // CHECK-NEXT:     return 0 + _d_i * j + i * _d_j;
+  // CHECK-NEXT: }
+
   // CHECK: clad::ValueAndPushforward<double, double> mem_fn_1_pushforward(double i, double j, SimpleFunctions1 *_d_this, double _d_i, double _d_j) {
   // CHECK-NEXT:     double _t0 = (this->x + this->y);
   // CHECK-NEXT:     double _t1 = i * j;
@@ -216,5 +233,4 @@ int main() {
   // CHECK: clad::ValueAndPushforward<SimpleFunctions1, SimpleFunctions1> operator_plus_pushforward(const SimpleFunctions1 &other, const SimpleFunctions1 *_d_this, const SimpleFunctions1 &_d_other) const {
   // CHECK-NEXT:     return {SimpleFunctions1(this->x + other.x, this->y + other.y), SimpleFunctions1(_d_this->x + _d_other.x, 0. + 0.)};
   // CHECK-NEXT: }
-
 }

--- a/test/Gradient/NonDifferentiable.C
+++ b/test/Gradient/NonDifferentiable.C
@@ -82,8 +82,17 @@ double fn_s2_operator(double i, double j) {
 }
 
 double fn_non_diff_var(double i, double j) {
-    non_differentiable double k = i * i * j;
-    return k;
+  non_differentiable double k = i * i * j;
+  return k;
+}
+
+non_differentiable
+double fn_non_diff(double i, double j) {
+  return i * j;
+}
+
+double fn_non_diff_call(double i, double j) {
+  return fn_non_diff(i, j) + i * j;
 }
 
 #define INIT_EXPR(classname)                                                   \
@@ -130,6 +139,8 @@ int main() {
   /*TEST_FUNC(fn_s2_operator, 3, 5)*/
 
   TEST_FUNC(fn_non_diff_var, 3, 5) // CHECK-EXEC: 0.00 0.00
+
+  TEST_FUNC(fn_non_diff_call, 3, 5) // CHECK-EXEC: 5.00 3.00
 
     // CHECK: void mem_fn_1_pullback(double i, double j, double _d_y, SimpleFunctions1 *_d_this, double *_d_i, double *_d_j);
 
@@ -181,6 +192,13 @@ int main() {
 
     // CHECK: void fn_non_diff_var_grad(double i, double j, double *_d_i, double *_d_j) {
     // CHECK-NEXT:     double k = i * i * j;
+    // CHECK-NEXT: }
+  
+    // CHECK: void fn_non_diff_call_grad(double i, double j, double *_d_i, double *_d_j) {
+    // CHECK-NEXT:     {
+    // CHECK-NEXT:         *_d_i += 1 * j;
+    // CHECK-NEXT:         *_d_j += i * 1;
+    // CHECK-NEXT:     }
     // CHECK-NEXT: }
     
     // CHECK: void mem_fn_1_pullback(double i, double j, double _d_y, SimpleFunctions1 *_d_this, double *_d_i, double *_d_j) {

